zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/ProcessInstanceMigrationClusteredTest.java

```diff
@@ -0,0 +1,411 @@
+          .endEvent()
+          .done();
+
+  private static final BpmnModelInstance TARGET_PROCESS =
+      Bpmn.createExecutableProcess("targetProcess")
+          .startEvent()
+          .userTask("B")
+          .boundaryEvent("boundary2")
+          .message(m -> m.name("message2").zeebeCorrelationKeyExpression("key2"))
+          .endEvent()
+          .moveToActivity("B")
+          .endEvent()
+          .done();
+
+  private static final BpmnModelInstance TARGET_PROCESS_2 =
+      Bpmn.createExecutableProcess("targetProcess2")
+          .startEvent()
+          .userTask("C")
+          .boundaryEvent("boundary3")
+          .message(m -> m.name("message3").zeebeCorrelationKeyExpression("key3"))
+          .endEvent()
+          .moveToActivity("C")
+          .endEvent()
+          .done();
+
+  public final Timeout testTimeout = Timeout.seconds(120);
+  public final ClusteringRule clusteringRule =
+      new ClusteringRule(
+          3,
+          1,
+          3,
+          brokerCfg ->
+              brokerCfg.getExperimental().getConsistencyChecks().setEnablePreconditions(true));
+  public final GrpcClientRule clientRule = new GrpcClientRule(clusteringRule);
+
+  @Rule
+  public RuleChain ruleChain =
+      RuleChain.outerRule(testTimeout).around(clusteringRule).around(clientRule);
+
+  @Test
+  public void shouldRejectMigrationWhenMessageSubscriptionMigrationIsNotDistributed() {
+    // given
+    final DeploymentEvent deploymentEvent =
+        clientRule
+            .getClient()
+            .newDeployResourceCommand()
+            .addProcessModel(SOURCE_PROCESS, "process.bpmn")
+            .addProcessModel(TARGET_PROCESS, "targetProcess.bpmn")
+            .addProcessModel(TARGET_PROCESS_2, "targetProcess2.bpmn")
+            .send()
+            .join();
+    final var sourceProcessDefinitionKey =
+        deploymentEvent.getProcesses().get(0).getProcessDefinitionKey();
+    final var targetProcessDefinitionKey =
+        deploymentEvent.getProcesses().get(1).getProcessDefinitionKey();
+    final var secondTargetProcessDefinitionKey =
+        deploymentEvent.getProcesses().get(2).getProcessDefinitionKey();
+
+    assertThat(
+            RecordingExporter.commandDistributionRecords(CommandDistributionIntent.ACKNOWLEDGED)
+                .withDistributionPartitionId(2)
+                .findFirst())
+        .isPresent();
+    assertThat(
+            RecordingExporter.commandDistributionRecords(CommandDistributionIntent.ACKNOWLEDGED)
+                .withDistributionPartitionId(3)
+                .findFirst())
+        .isPresent();
+
+    int processInstancePartitionId = 0;
+    int subscriptionPartitionId = 0;
+    long processInstanceKey = -1;
+
+    // To test multi partition behaviour of the migration, we need to ensure that the process
+    // instance and the message subscription are on different partitions
+    // TODO - make process instance partition id deterministic, so no need to loop
```

Comment:
@korthout I am looking for a simpler way to make sure process instance partition id and subscription partition id are different. Is there any?

Reply:
You can create a process instance on a specific partition using: `clusteringRule.createProcessInstanceOnPartition()`. Just pick one by hand and hardcode it. Same goes for the correlation key, just choose one that results in a specific partition id. Lastly, ensure that the preconditions for the test are correct, by asserting that the partitions differ.

Reply:
That is way more clear, updated! 🚀